doc/queue-sheet.1
.\" Automatically generated by Pandoc 3.0.1
.\"
.\" Define V font for inline verbatim, using C font in formats
.\" that render this, and otherwise B font.
.ie "\f[CB]x\f[]"x" \{\
. ftr V B
. ftr VI BI
. ftr VB B
. ftr VBI BI
.\}
.el \{\
. ftr V CR
. ftr VI CI
. ftr VB CB
. ftr VBI CBI
.\}
.TH "QUEUE-SHEET" "1" "" "queue-sheet-haskell 0.8.0.1 (2026-01-01)" "queue-sheet Manual"
.nh
.SH NAME
.PP
\f[V]queue-sheet\f[R] - queue sheet utility
.SH SYNOPSIS
.PP
\f[V]queue-sheet\f[R] [\f[I]OPTIONS\f[R]] QUEUES.yaml
.SH DESCRIPTION
.PP
Queue Sheet is a utility that builds PDFs of lists.
Printed PDFs can be used to track progress through queues when offline.
.SH OPTIONS
.TP
-h, --help
show help and exit
.TP
--version
show version and exit
.TP
-t, --template \f[I]TEMPLATE.tex\f[R]
template file (default: template.tex)
.TP
-o, --output \f[I]QUEUES.pdf\f[R]
output file
.SH ARGUMENTS
.TP
\f[I]QUEUES.yaml\f[R]
queue file
.SH FILES
.SS \f[V]QUEUES.yaml\f[R]
.PP
A queue is a named list of items.
For example, a podcast can be represented as a queue where the queue
name is the name of the podcast and each item in the queue is an episode
of the podcast.
Queues can optionally be organized into sections.
For example, sections can be used to organize podcast queues by theme.
.PP
Queues are specified in YAML format.
They may be specified in a few different ways, depending on how you want
to organize them.
.PP
To create a queue sheet of queues without sections, the YAML file
consists of an array of queue objects, which have the following
properties:
.TP
\f[I]name\f[R]
queue name (string, required)
.TP
\f[I]url\f[R]
queue URL (string, optional)
.TP
\f[I]date\f[R]
date of last update (string, optional)
.TP
\f[I]tags\f[R]
list of tags (space-separated string or array of strings, optional)
.TP
\f[I]prev\f[R]
previous item (item, optional)
.TP
\f[I]next\f[R]
list of next items (space-separated string or array of items, optional)
.PP
The only required property is \f[I]name\f[R].
.PP
The \f[I]tags\f[R] property associates one or more string tags with the
queue.
Tags can be specified using a space-separated string or an array of
strings.
A tag must consist of at least one ASCII letter, number, period,
underscore, or dash.
For example, tag \[lq]complete\[rq] can be used to indicate that there
will be no new episodes of a podcast that is complete.
.PP
The \f[I]next\f[R] property is a list of next items in the queue.
When the list is exhausted, the previous item can be specified using the
\f[I]prev\f[R] property.
If both \f[I]prev\f[R] and \f[I]next\f[R] are specified, \f[I]prev\f[R]
is ignored.
.PP
Next items can be specified using a space-separated string or an array
or items.
When specifying items using an array, the item name can be specified
using a string or a number, or an object with the following properties
can be used in order to specify more item information:
.TP
\f[I]name\f[R]
name of the item (string, required)
.TP
\f[I]url\f[R]
item URL (string, optional)
.TP
\f[I]tags\f[R]
list of tags (space-separated string or array of strings, optional)
.PP
To organize queues into sections, the YAML file should be written as an
object with two properties:
.TP
\f[I]sections\f[R]
list of sections names (array of strings, optional)
.TP
\f[I]queues\f[R]
list of queue objects (array of queue objects, required)
.PP
Sections names are specified using strings.
The order that the sections are specified determines the order that they
are displayed on the queue sheet.
.PP
Queue objects are as above, with an additional property to specify the
section:
.TP
\f[I]section\f[R]
name of the section (string, optional)
.PP
Queues that are not explicitly associated with a section are associated
with an implicit default section.
.PP
To make it easier to share queue files, imports are also supported.
Import another queue file using an import object instead of a queue
object in an array of queues.
An import object has the following properties:
.TP
\f[I]import\f[R]
path to the queue file to import (string, required)
.TP
\f[I]section\f[R]
section to associate all of the imported queues with (string, optional)
.PP
Paths are relative to the current file.
For example, simply specify the filename of the file to import when the
file is in the same directory.
.PP
When you specify \f[I]section\f[R], the section must be defined in the
current file.
When you do not specify \f[I]section\f[R], the sections of the imported
queues are used, but they must also be defined in the current file.
.SS Template
.PP
YAML files specify the data, and templates determine how that data is
displayed.
A LaTeX template is used to build the PDF, using XeTeX.
Unless specified otherwise, \f[V]template.tex\f[R] is used.
.PP
It is a Jinja2-style template using the following syntax:
.TP
Interpolations
\f[V]<< section.name >>\f[R]
.TP
Tags
\f[V]<! if section.name !>\f[R]
.TP
Comments
\f[V]<# comment #>\f[R]
.PP
The context contains a single value:
.TP
\f[I]sections\f[R]
list of sections
.PP
A section is an object with the following properties:
.TP
\f[I]name\f[R]
name of the section (string)
.TP
\f[I]queues\f[R]
list of queues
.PP
Only sections that contain queues are passed to the template.
The \f[I]name\f[R] property of the default section is empty.
.PP
A queue is an object with the following properties:
.TP
\f[I]name\f[R]
name of the queue (string)
.TP
\f[I]url\f[R]
queue URL or empty string if no URL (string)
.TP
\f[I]date\f[R]
date or empty string if no date (string)
.TP
\f[I]prev_item\f[R]
previous item or empty string if not set (item)
.TP
\f[I]next_items\f[R]
list of next items (list of items)
.PP
Queue tags are exposed as boolean properties prefixed with
\[lq]tag_\[rq].
For example, a tag named \[lq]complete\[rq] is exposed as
\[lq]tag_complete\[rq].
.PP
An item is an object with the following properties:
.TP
\f[I]name\f[R]
name of the item (string)
.TP
\f[I]url\f[R]
item URL or empty string if no URL (string)
.PP
Item tags are exposed as boolean properties prefixed with
\[lq]tag_\[rq].
For example, a tag named \[lq]em\[rq] is exposed as \[lq]tag_em\[rq].
.SS \f[V]QUEUES.pdf\f[R]
.PP
Unless specified otherwise, the built PDF is output to a file with the
same base name as the queues file but with a \f[V].pdf\f[R] extension.
.SH EXIT CODES
.TP
0
no error
.TP
1
execution error
.TP
2
command-line error
.SH PROJECT
.TP
GitHub:
<https://github.com/ExtremaIS/queue-sheet-haskell>
.TP
Reporting issues:
GitHub: <https://github.com/ExtremaIS/queue-sheet-haskell/issues>
.RS
.PP
Email: <bugs@extrema.is>
.RE
.TP
Copyright
Copyright (c) 2020-2026 Travis Cardwell
.TP
License
The MIT License <https://opensource.org/licenses/MIT>


CHANGELOG.md
# `queue-sheet-haskell` Changelog

This project follows the [Haskell package versioning policy][PVP], with
versions in `A.B.C.D` format.  `A` may be incremented arbitrarily for
non-technical reasons, but [semantic versioning][SemVer] is otherwise
followed, where `A.B` is the major version, `C` is the minor version, and `D`
is the patch version.  Initial development uses versions `0.0.0.D`, for which
every version is considered breaking.

[PVP]: <https://pvp.haskell.org/>
[SemVer]: <https://semver.org/>

The format of this changelog is based on [Keep a Changelog][KaC], with the
following conventions:

* Level-two heading `Unreleased` is used to track changes that have not been
  released.
* Other level-two headings specify the release in `A.B.C.D (YYYY-MM-DD)`
  format, with newer versions above older versions.
* Level-three headings are used to categorize changes as follows:
    1. Breaking
    2. Non-Breaking
* Changes are listed in arbitrary order and present tense.

[KaC]: <https://keepachangelog.com/en/1.0.0/>

## Unreleased

### Non-Breaking

* Bump `optparse-applicative` dependency version upper bound

## 0.8.0.1 (2025-01-04)

### Non-Breaking

* Flip `optparse-applicative_ge_0_18` flag
* Bump `base` dependency version upper bound

## 0.8.0.0 (2025-01-03)

### Breaking

* Support whitespace-separated tags and items instead of CSV
* Remove support for GHC 8.6, constraining lower bounds
* Remove support for GHC 8.4, constraining lower bounds
* Remove support for GHC 8.2, constraining lower bounds
* Change minimal Cabal from 1.24 to 3.0

### Non-Breaking

* Bump `aeson` dependency version upper bound
* Bump `ansi-wl-pprint` dependency version upper bound
* Bump `base` dependency version upper bound
* Bump `bytestring` dependency version upper bound
* Bump `filepath` dependency version upper bound
* Add support for `optparse-applicative` `0.18`
* Bump `tasty` dependency version upper bound
* Bump `text` dependency version upper bound
* Bump `transformers` dependency version upper bound
* Bump `ttc` dependency version upper bound
* Bump `vector` dependency version upper bound
* Adjust dependency constraints to match tested versions

## 0.7.0.2 (2022-03-02)

### Non-Breaking

* Bump `text` dependency version upper bound
* Bump `optparse-applicative` dependency version upper bound

## 0.7.0.1 (2021-10-10)

### Non-Breaking

* Add dependency bounds

## 0.7.0.0 (2021-06-25)

### Breaking

* Fix `--help` when using `optparse-applicative` `0.16`

### Non-Breaking

* Use TTC 1.1.0.1
* Refactor Nix configuration

## 0.6.0.0 (2021-05-27)

### Breaking

* Add support for `optparse-applicative` `0.16`

### Non-Breaking

* Add `.deb` and `.rpm` packaging
* Add Cabal support to `Makefile`
* Add Cabal tests to GitHub Actions
* Add [stan](https://hackage.haskell.org/package/stan) static analysis

## 0.5.0.1 (2020-11-23)

### Non-Breaking

* Use GitHub Actions instead of Travis CI

## 0.5.0.0 (2020-11-08)

### Breaking

* Add simplified CSV support for tags and items

### Non-Breaking

* Rename Git default branch to `main`

## 0.4.0.0 (2020-09-12)

### Breaking

* Add item support for tags

## 0.3.0.0 (2020-08-23)

### Breaking

* Add default section support
* Add top-level queues array support
* Add import support
* Add user-defined tag support
* Remove split property
* Change templates to use snake-case properties

## 0.2.0.0 (2020-08-11)

### Breaking

* Add URL support

## 0.1.0.0 (2020-07-25)

### Breaking

* Configure sections and queues in a single YAML file

### Non-Breaking

* Add a template option
* Add an output option
* Refactor `Makefile`, add `STACK_NIX_PATH` support
* Add `test-all` command to `Makefile`
* Add Nix configuration

## 0.0.1.0 (2020-01-22)

### Breaking

* Initial public release


app/queue-sheet.hs
------------------------------------------------------------------------------
-- |
-- Module      : Main
-- Description : queue-sheet program
-- Copyright   : Copyright (c) 2020-2026 Travis Cardwell
-- License     : MIT
------------------------------------------------------------------------------

{-# LANGUAGE RecordWildCards #-}

module Main (main) where

-- https://hackage.haskell.org/package/base
import Control.Applicative (optional)
import System.Exit (exitFailure)

-- https://hackage.haskell.org/package/optparse-applicative
import qualified Options.Applicative as OA

-- (queue-sheet)
import QueueSheet (version)
import QueueSheet.Build (buildPdf)

-- (queue-sheet:executable)
import qualified LibOA

------------------------------------------------------------------------------
-- $Constants

-- | Default template file
defaultTemplate :: FilePath
defaultTemplate = "template.tex"

------------------------------------------------------------------------------
-- $Library

-- | Display an error and exit the program
errorExit :: String -> IO a
errorExit msg = do
    putStrLn $ "error: " ++ msg
    exitFailure

------------------------------------------------------------------------------
-- $Options

-- | Program options
data Options
  = Options
    { optTemplate :: !FilePath
    , optOutput   :: !(Maybe FilePath)
    , optQueues   :: !FilePath
    }
  deriving Show

-- | Parse program options
parseOptions :: IO Options
parseOptions = OA.execParser
    $ OA.info (LibOA.helper <*> LibOA.versioner version <*> options)
    $ mconcat
        [ OA.fullDesc
        , OA.progDesc "queue sheet utility"
        , OA.failureCode 2
        ]
  where
    options :: OA.Parser Options
    options = Options
      <$> templateOption
      <*> optional outputOption
      <*> queuesArgument

    templateOption :: OA.Parser FilePath
    templateOption = OA.strOption $ mconcat
      [ OA.long "template"
      , OA.short 't'
      , OA.metavar "TEMPLATE.tex"
      , OA.value defaultTemplate
      , OA.showDefaultWith id
      , OA.help "template file"
      ]

    outputOption :: OA.Parser FilePath
    outputOption = OA.strOption $ mconcat
      [ OA.long "output"
      , OA.short 'o'
      , OA.metavar "QUEUES.pdf"
      , OA.help "output file"
      ]

    queuesArgument :: OA.Parser FilePath
    queuesArgument = OA.strArgument $ mconcat
      [ OA.metavar "QUEUES.yaml"
      , OA.help "YAML file specifying queue information"
      ]

------------------------------------------------------------------------------
-- $main

main :: IO ()
main = do
  Options{..} <- parseOptions
  either errorExit return =<< buildPdf optQueues optTemplate optOutput
